benchmark-overhead/README.md

```diff
@@ -1,27 +1,107 @@
+For each agent defined in a configuration, the test runner (using [testcontainers](https://www.testcontainers.org/)) will:
+1. create a fresh postgres instance and populate it with initial data.
+2. create a fresh instance of [spring-petclinic-rest](https://github.com/spring-petclinic/spring-petclinic-rest) instrumented with the specified agent
+3. measure the time until the petclinic app is marked "healthy" and then write it to a file
+4. if configured, perform a warmup phase. During the warmup phase, a bit of traffic is generated in order to get the application into a steady state (primarily helping facilitate jit compilations). Currently, we use a 30 second warmup time. 
+5. start a JFR recording by running `jcmd` inside the petclinic container
+6. run the [k6 test script](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/benchmark-overhead/k6/basic.js) with the configured number of iterations through the file and the configured number of concurrent virtual users (VUs).
+7. after k6 completes, petclinic is shut down
+8. after petclinic is shut down, postgres is shut down
+
+And this repeats for every agent configured in each test configuration.
+
+After all the tests are complete, the results are collected and commited back to the `/results` subdirectory as csv and summary text files.
+
+## What do we measure?
+
+For each test pass, we record the following metrics in order to compare agents and determine
+relative overhead.
+
+| metric name              | description                                                     |
+|--------------------------|-----------------------------------------------------------------|
+| Startup time       | how long it takes for the spring app to report "healthy"
+| Total allocated MB | across the life of the application
+| Heap (min)         | smallest observed heap size
+| Heap (max)         | largest observed heap size
+| Thread switch rate | max observed thread context switch rate
+| GC time            | total amount of time spent paused for garbage collection
+| Request mean       | average time to handle a single web request (measured at the caller)
+| Request p95        | 95th percentile time to handle a single web requ4st (measured at the caller)
+| Iteration mean     | average time to do a single pass through the k6 test script
+| Iteration p95      | 95th percentile time to do a single pass through the k6 test script
+| Peak threads       | Highest number of running threads in the VM, including agent threads
+| Network read mean  | Average network read rate
+| Network write mean | Average network write rate
 
 ## Config
 
-tbd
+Each config contains the following:
+* name
+* description
+* list of agents (see below)
+* maxRequestRate (optional, used to throttle traffic)
+* concurrentConnections (number of concurrent virtual users [VUs])
+* totalIterations - the number of passes to make through the k6 test script
```

Comment:
These are after warm up, right?

Reply:
Yes, after warmup. The only thing not after warmup is startup time.